README.md
# NMOS Registry Registration API Service

Provides a service implementing the NMOS Registration API, specified in AMWA IS-04.

The API will present itself at [http://localhost:8235/x-nmos/registration/].

## Installing with Python

Before installing this library please make sure you have installed the [NMOS Common Library](https://github.com/bbc/nmos-common), on which this API depends.

The registration API also requires a datastore. Either [etcd](https://github.com/coreos/etcd) or [Couchbase Server](https://www.couchbase.com) to be installed.

The legacy etcd store can be installed on debain distributions using apt:

```
sudo apt-get install etcd

```

The modern Couchbase datastore [requires some further steps](https://docs.couchbase.com/python-sdk/2.5/start-using-sdk.html). The Python package depends on the C `libcouchbase` SDK, installed on debian distributions as follows:

```
# Only needed during first-time setup:
wget http://packages.couchbase.com/releases/couchbase-release/couchbase-release-1.0-6-amd64.deb
sudo dpkg -i couchbase-release-1.0-6-amd64.deb
# Will install or upgrade packages
sudo apt-get update
sudo apt-get install libcouchbase-dev libcouchbase2-bin build-essential
```

In addition `libsystemd-dev` is required to  be installed for `cysystemd`:

```
sudo apt install libsystemd-dev
```

Once all dependencies are satisfied run the following commands to install the API:

```
pip install setuptools
sudo python setup.py install
```

## Configuration

The Registration API makes use of two configuration files. The first is native to the Registration API and is described below. The second forms part of the [NMOS Common Library](https://github.com/bbc/nmos-common) and is described in that repository. Note that you will likely have to configure items in both files.

The native Registration API configuration should consist of a JSON object in the file `/etc/ips-regaggregator/config.json`. The following attributes may be set within the object:

*   **priority:** \[integer\] Sets a priority value for this Registration API instance between 0 and 255. A value of 100+ indicates a development rather than production instance. Default: 100.
*   **https_mode:** \[string\] Switches the API between HTTP and HTTPS operation. "disabled" indicates HTTP mode is in use, "enabled" indicates HTTPS mode is in use. Default: "disabled".
*   **enable_mdns:** \[boolean\] Provides a mechanism to disable mDNS announcements in an environment where unicast DNS is preferred. Default: true.
*   **oauth_mode:** \[boolean\] Switches the API between being secured using OAuth2 and not using authorization. Default: false.

An example configuration file is shown below:

```json
{
  "priority": "30",
  "https_mode": "enabled",
  "enable_mdns": false,
  "oauth_mode": true
}
```

## Running the Registration API

### Non-blocking

Run the following script to start the Registration API in a non-blocking manner, and then stop it again at a later point:

```Python
    from nmosregistration.registryaggregatorservice import RegistryAggregatorService

    service = RegistryAggregatorService()
    service.start()

    # Do something else until ready to stop

    service.stop()
```

### Blocking

It is also possible to run Registration API in a blocking manner:

```python
from nmosregistration.registryaggregatorservice import RegistryAggregatorService

service = RegistryAggregatorService()
service.run() # Runs forever
```

### Local Development Datastore

To run a local instance of Couchbase Server, the easiest approach is to run `docker-compose` in the `tests/` directory and run a test container. Some setup is required, either through a web GUI accessible via `http://[host]:8091` or POSTing the http requests specified in `_initialise_cluster()` within `test_couchbase` in `tests/v1_0/`.

For a full development environment, a `Vagrantfile` is provided in the root of the directory that provisions a Virtual Machine with all the required dependencies. This can be run using:

```bash
vagrant up --provision
```

Note this will require both Vagrant and Ansible to be installed in order to provision the VM. Once the machine has been provisioned, running the following script will configure a single node cluster and bring up the API Service:

```bash
vagrant ssh  # ssh into the VM
python3 /vagrant/initialise_couchbase.py  # initialise cluster
```

## Tests

Unit tests are provided.  Currently these have hard-coded dummy/example hostnames, IP addresses and UUIDs.  You will need to edit the Python files in the test/ directories to suit your needs and then "make test".

## Debian Packaging

Debian packaging files are provided for internal BBC R&D use.
These packages depend on packages only available from BBC R&D internal mirrors, and will not build in other environments. For use outside the BBC please use python installation method.


CHANGELOG.md
# NMOS Registration API Implementation Changelog

## 0.8.1
- Replace RequiresAuth decorator with AuthMiddleware middleware

## 0.8.0
- Use official etcd ports

## 0.7.11
- Alter executable to run using Python3, alter `stdeb` to replace python 2 package

## 0.7.10
- Clean-up before stopping service thread

## 0.7.9
- Add `api_auth` text record to multicast announcements

## 0.7.8
- Update schemas for v1.3 and earlier changes

## 0.7.7
- Import config from seperate file, add OAUTH_MODE config parameter

## 0.7.6
- Add cleanup function when stopping service

## 0.7.5
- Fix python3 errors

## 0.7.4
- Move NMOS packages from recommends to depends

## 0.7.3
- Add systemd ready notification when service has started

## 0.7.2
- Add Python3 linting stage to CI, fix linting

## 0.7.1
- Fix missing files in Python 3 Debian package

## 0.7.0
- Addition of OAuth2 security decorators and added linting stage to Jenkins CI

## 0.6.6
- Made to work with python 3

## 0.6.5
- Add missing dependency

## 0.6.4
- Fix bug causing format validation to be skipped

## 0.6.3
- Fix bug preventing use of priorities 1 through 99

## 0.6.2
- Update mDNS behaviour based on latest v1.3 draft

## 0.6.1
- Update schemas for v1.3

## 0.6.0
- Add support for new mDNS service type in v1.3

## 0.5.0
- Add config option to enable/disable mDNS announcement

## 0.4.0
- Disable v1.0 API when running in HTTPS mode

## 0.3.0
- Add provisional support for IS-04 v1.3


nmosregistration/config.py
# Copyright 2017 British Broadcasting Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import absolute_import
from __future__ import print_function
import os
import json
import copy

# HTTPS under test only at present
# enabled = Use HTTPS only in all URLs and mDNS adverts
# disabled = Use HTTP only in all URLs and mDNS adverts
# mixed = Use HTTP in all URLs, but additionally advertise an HTTPS endpoint for discovery of this API only
CONFIG_DEFAULTS = {
    "priority": 100,
    "https_mode": "disabled",
    "enable_mdns": True,
    "oauth_mode": False,
    "service_port": 8235,
    "registry": {
        "type": "couchbase",
        "hosts": [""],
        "port": 8091,
        "username": "",
        "password": "",
        "buckets": {
            "registry": "",
            "meta": ""
        }
    },
    "resource_expiry": 12,
    "ws_period": 30
}

config = {}
config.update(copy.deepcopy(CONFIG_DEFAULTS))

try:
    # Check for current nmos config file and legacy ipstudio file
    nmos_config_file = "/etc/nmos-registration/config.json"
    ipstudio_config_file = "/etc/ips-regaggregator/config.json"
    if os.path.isfile(nmos_config_file):
        f = open(nmos_config_file, 'r')
    elif os.path.isfile(ipstudio_config_file):
        f = open(ipstudio_config_file, 'r')
    if f:
        extra_config = json.loads(f.read())
        config.update(extra_config)
except Exception as e:
    print("Exception loading config: {}".format(e))
